File: card-lang/src/calculator/parser/symbols/identifier.rs
use std::collections::HashMap;

use crate::types::Parser::Parser;
use crate::types::Parser::NudListener;
use crate::types::Parser::SymbolAndToken;
use crate::types::ReusedStructs::Node;
use crate::types::ReusedStructs::NodeType;
use crate::types::ReusedStructs::Token;


pub struct IdentifierNud {}

impl NudListener for IdentifierNud {
  fn run(&self, symtok: SymbolAndToken, parser: Parser)->Node{
    let token = symtok.token.clone();
    if !isOpener(symtok.token) {
      return Node {
        node_type: NodeType::IdentifierNode,
        values: Some(token.values),
        branches: None,
        args: None
      }
    }

    let mut args: Vec<Node> = vec![];
    let (p, t) = parser.nextToken();
    let mut p1 = p;
    let mut n:Node;
    if isCloser(t) {
      (p1, _) = p1.advance();
    } else {
      loop {
        (p1, _) = p1.advance();
        (p1, n) = p1.expression(2);
        args.push(n);
        let t: SymbolAndToken;
        (p1, t) = p1.token();
        if!isComma(t.token) { break }
      }
      let t: SymbolAndToken;
      (p1, t) = p1.token();
      if !isCloser(t.token) {
        panic!("Expected closing parenthesis ')'");
      };
    }
    p1.advance();
    return Node {
      node_type: NodeType::CallNode,
      values: Some(token.values),
      args: Some(args),
      branches: None,
    }
  }
}

fn isOpener(token: Token) -> bool {
  if token.token_type != "operator" {
    return false
  }
  let str: String = token.values.get("value").unwrap().clone();
  if  str != "(".to_string() {
    return false;
  }
  return true;
}

fn isCloser(token: Token) -> bool {
  if token.token_type != "operator" {
    return false
  }
  let str = token.values.get("value").unwrap().clone();
  if  str != ")".to_string() {
    return false;
  }
  return true;
}

fn isComma(token: Token) -> bool {
  if token.token_type != "operator" {
    return false
  }
  let str = token.values.get("value").unwrap().clone();
  if  str != ",".to_string() {
    return false;
  }
  return true;

}

File: card-lang-typescript/src/calculator/parser/symbols/identifier.ts
import { NudListener, ParserController, SymbolAndToken } from "../../../types/Parser";
import { LangNode, LangNodeType, Token } from "../../../types/ReusedStructs";

export class IdentifierNud implements NudListener {
  run(symtok: SymbolAndToken, parser: ParserController): LangNode{
    let token = symtok.token;
    let name = token.values?.name;
    if(name == void 0){
      throw new Error("Needed a name for identifier");
    }
    // parser.advance()
    var nextTok = parser.token();
    if(typeof nextTok == "undefined"){
      return {
        node_type: LangNodeType.IdentifierNode,
        values: { name },
      }
    }
    if(!isOpener(nextTok.token)){
      return {
        node_type: LangNodeType.IdentifierNode,
        values: { name },
      }
    }

    var args: Array<LangNode> = [];
    let t: void | Token = parser.nextToken();
    if(typeof t == "undefined"){
      throw new Error("expected a token, reached end of token list");
    }

    if(isCloser(t)){
      parser.advance();
    } else {
      while(true){
        parser.advance();
        args.push(parser.expression(2));
        let s = parser.token();
        if(typeof s == "undefined"){
          throw new Error("expected token, reached end of list");
        }
        if(!isComma(s.token)){ break }
      }
      let s = parser.token();
      if(typeof s == "undefined"){
        throw new Error("expected closer, reached end of list");
      }
      if(!isCloser(s.token)){
        var value = s.token.values.value;
        throw new Error(`Expected closing parenthesis ')' got ${value}`);
      };
    }
    parser.advance();
    return {
      node_type: LangNodeType.CallNode,
      values: { name },
      args: args,
    }
  }
}

function isOpener(token: Token): boolean {
  if(token.token_type != "operator"){
    return false
  }
  let str: String = token.values.value;
  if(str != "("){
    return false;
  }
  return true;
}

function isCloser(token: Token): boolean {
  if(token.token_type != "operator"){
    return false
  }
  let str = token.values.value;
  if(str != ")"){
    return false;
  }
  return true;
}

function isComma(token: Token): boolean {
  if(token.token_type != "operator"){
    return false
  }
  let str = token.values.value;
  if(str != ","){
    return false;
  }
  return true;

}
